File: src/modules/graphics/graphics.cpp
#include <raylib.h>
#include "graphics.hpp"

namespace starry {

void graphics::clear(Color color)
{
    BeginDrawing();
    ClearBackground({ color.r, color.g, color.b, color.a });
}

void graphics::end_drawing()
{
    EndDrawing();
}

void graphics::draw_texture(Texture texture, Vec2 pos, float32 angle)
{
    graphics::draw_texture_ext(
        texture,
        Vec2(0, 0),
        Vec2(texture.size.x, texture.size.y),
        pos,
        Vec2(texture.size.x, texture.size.y),
        Vec2(0.5, 0.5),
        angle,
        ST_COLOR_WHITE
    );
}

void graphics::draw_texture_ext(Texture texture, Vec2 src_pos, Vec2 src_size, Vec2 dst_pos,
    Vec2 dst_size, Vec2 origin, float32 rotation, Color tint)
{
    // akson dendryt
    DrawTexturePro(
        {
            texture.__data.id,
            texture.__data.width,
            texture.__data.height,
            texture.__data.mipmaps,
            texture.__data.format,
        },
        {(float)src_pos.x, (float)src_pos.y, (float)src_size.x, (float)src_size.y},
        {(float)dst_pos.x, (float)dst_pos.y, (float)dst_size.x, (float)dst_size.y},
        {(float)origin.x * (float)dst_size.x, (float)origin.y * (float)dst_size.y},
        rotation,
        {tint.r, tint.g, tint.b, tint.a}
    );
}

}

File: modules/graphics/graphics.go
// useful for when you're not blind
package graphics

import (
	"image/color"

	rl "github.com/gen2brain/raylib-go/raylib"
	"github.com/hellory4n/stellarthing/core"
)

// clears the screen
func Clear(color core.Color) {
	rl.BeginDrawing()
	rl.ClearBackground(rl.NewColor(color.R, color.G, color.B, color.A))
}

// as the name implies, it ends drawing
func EndDrawing() {
	rl.EndDrawing()
}

// this is utter insanity. srcPos and srcSize are used for cropping the image, dstPos and dstSize are the actual position and size, origin is a multiplier so (0, 0) is the top left, (0.5, 0.5) is the center, and (1, 1) is the bottom right. everything else is self-explanatory
func DrawTextureExt(texture Texture, srcPos core.Vec2, srcSize core.Vec2, dstPos core.Vec2, dstSize core.Vec2, origin core.Vec2, angle float64, tint core.Color) {
	rl.DrawTexturePro(
		rl.Texture2D{
			ID: texture.id,
			Width: texture.width,
			Height: texture.height,
			Format: texture.format,
			Mipmaps: texture.mipmaps,
		},
		rl.Rectangle{X: float32(srcPos.X), Y:  float32(srcPos.Y), Width: float32(srcSize.X), Height: float32(srcSize.Y)},
		rl.Rectangle{X: float32(dstPos.X), Y:  float32(dstPos.Y), Width: float32(dstSize.X), Height: float32(dstSize.Y)},
		rl.Vector2{X: float32(origin.X * dstSize.X), Y: float32(origin.Y * dstSize.Y)},
		float32(angle),
		color.RGBA{R: tint.R, G: tint.G, B: tint.B, A: tint.A},
	)
}

// draws a texture. if you want a comically large parameter list use DrawTextureExt instead
func DrawTexture(texture Texture, pos core.Vec2, angle float64, color core.Color) {
	DrawTextureExt(
		texture, core.NewVec2(0, 0), texture.Size().ToVec2(), pos, texture.Size().ToVec2(),
		core.NewVec2(0.5, 0.5), angle, color,
	)
}